servant-client/src/Servant/Client/HttpClient.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DeriveDataTypeable         #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+
+{-| http-client based client  requests executor -}
+module Servant.Client.HttpClient where
+
+
+import Prelude ()
+import Prelude.Compat
+
+import Control.Exception
+import Control.Monad
+import Control.Monad.Catch (MonadThrow, MonadCatch)
+import Data.Foldable (toList)
+import Data.Functor.Alt (Alt (..))
+
+import Control.Monad.Error.Class (MonadError(..))
+import Control.Monad.Trans.Except
+
+import GHC.Generics
+import Control.Monad.Base (MonadBase (..))
+import Control.Monad.IO.Class ()
+import Control.Monad.Reader
+import Control.Monad.Trans.Control (MonadBaseControl (..))
+import Data.ByteString.Lazy hiding (pack, filter, map, null, elem, any)
+import Data.String.Conversions (cs)
+import Data.Proxy
+import Network.HTTP.Media
+import Network.HTTP.Types
+import Network.HTTP.Client hiding (Proxy, path)
+import qualified Network.HTTP.Types.Header   as HTTP
+import Servant.API.ContentTypes
+import Servant.Client.Class
+import Servant.Common.BaseUrl
+import Servant.Common.Req
+
+import qualified Network.HTTP.Client as Client
+
+instance RunClient ClientM NoContent ( Int, ByteString, MediaType
+                                     , [HTTP.Header], Response ByteString) where
+  runRequest _ meth req = performRequest meth req
+
+instance (MimeUnrender ct a) =>
```

Comment:
Looking at this again, it seems a little obscure how the types determine or correspond with behavior. How would a person know that these are the instances they have to write, and with this behavior? Since `RunClient` will be part of the external API, this question matters.

Is it not sufficient to have `runRequest` have type:

```
  runRequest :: Proxy ct
             -> Method
             -> Req
             -> m ( Int, ByteString, MediaType, [HTTP.Header], Response ByteString)
```
And then define `runRequestCT` and `runRequestNoBody` in terms of these? Or if that doesn't work, to have a class with three methods (with maybe a `runRequest` `MINIMAL`)?